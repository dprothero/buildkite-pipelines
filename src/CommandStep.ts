import { CommandStepObject } from "./CommandStepObject";
import { PluginBuilder } from "./PluginBuilder";
import { StepBuilder } from "./StepBuilder";

export interface CommandStepOptions {
  command?: string | string[]
  label?: string;
  branches?: string[];
  plugins?: PluginBuilder[]

}

export class CommandStep implements StepBuilder {
  #key?: string;
  #dependsOn: StepBuilder[] = [];
  #label?: string;
  #branches: string[] = [];
  #command: string[]
  #plugins: PluginBuilder[] = []
  #artifactPaths: string[] = []

  public constructor(command: string | string[] | CommandStepOptions) {
    if (Array.isArray(command)) {
      this.#command = command 
    } else if (typeof command === 'string') {
      this.#command = [command]
    } else if (Array.isArray(command.command)) {
      this.#command = command.command
    } else {
      if (typeof command.command === 'string') {
        this.#command = [command.command]
      } else {
        this.#command = []
      }
      this.#label = command.label
      this.#plugins = command.plugins ?? []
    }
  }

  public get key(): string | undefined {
    return this.#key;
  }

  public set key(key: string | undefined) {
    this.#key = key;
  }

  public get dependsOn(): StepBuilder[] {
    return this.#dependsOn;
  }

  public dependOn(step: StepBuilder): this {
    this.#dependsOn?.push(step);
    return this;
  }

  public get label(): string | undefined {
    return this.#label;
  }

  public set label(label: string | undefined) {
    this.#label = label;
  }

  public get branches(): string[] {
    return this.#branches ?? [];
  }

  public addBranch(branch: string): this {
    this.#branches.push(branch);
    return this;
  }

  public get plugins(): PluginBuilder[] {
    return this.#plugins;
  }

  public addPlugin(builder: PluginBuilder): this {
    this.#plugins.push(builder)
    return this;
  }

  public get artifactPaths(): string[] {
    return this.#artifactPaths;
  }

  public addArtifactPath(path: string): this {
    this.#artifactPaths.push(path);
    return this;
  }

  public build(): CommandStepObject {
    return {
      key: this.#key,
      depends_on: this.#dependsOn.length ? this.#dependsOn.map(step => step.key) : undefined, 
      label: this.#label,
      commands: this.#command,
      plugins: this.#plugins.length ? this.#plugins.map((builder) => builder.build()) : undefined
    }
  }
}