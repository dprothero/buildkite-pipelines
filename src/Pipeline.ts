import { PipelineBuilder } from "./PipelineBuilder";
import { PipelineObject } from "./PipelineObject";
import { StepBuilder } from "./StepBuilder";

export interface PipelineOptions {
  steps?: StepBuilder[]
}

export class Pipeline implements PipelineBuilder {
  #steps: StepBuilder[];

  public constructor({steps}: PipelineOptions = {}) {
    this.#steps = steps ?? [];
  } 

  public build(): PipelineObject {
    return {
      steps: this.#steps.map(step => step.build())
    }
  }
}