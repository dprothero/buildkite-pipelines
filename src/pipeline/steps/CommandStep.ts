import { CommandStepObject } from "./CommandStepObject";
import { PluginBuilder } from "./plugins/PluginBuilder";
import { KeyMixin, KeyMixinMethods } from "./KeyMixin";
import { LabelMixin, LabelMixinMethods } from "./LabelMixin";
import { StepBuilder } from "./StepBuilder";
import { PluginObject } from "./plugins/PluginObject";
import { BranchesMixin, BranchesMixinMethods } from "./BranchesMixin";
import { ConditionalMixin, ConditionalMixinMethods } from "./ConditionalMixin";
import {
  DependenciesMixin,
  DependenciesMixinMethods,
} from "./DependenciesMixin";

function isBuilder(step: PluginObject | PluginBuilder): step is PluginBuilder {
  return typeof (step as any).build == "function";
}

export namespace CommandStep {
  export interface Builder
    extends StepBuilder,
      KeyMixinMethods,
      LabelMixinMethods,
      ConditionalMixinMethods,
      BranchesMixinMethods,
      DependenciesMixinMethods {
    command(command: string | string[]): this;
    env(env: Record<string, string | number>): this;
    addEnv(name: string, value: string): this;
    plugins(plugins: Array<PluginObject | PluginBuilder>): this;
    addPlugin(plugin: PluginObject | PluginBuilder): this;
    concurrency(jobs: number, group: string): this;
    artifactPaths(paths: string[]): this;
    addArtifactPath(path: string): this;
    timeout(minutes: number): this;
    softFail(fail: boolean): this;
    parallelism(count: number): this;
  }
}

export class CommandStep {
  static builder(): CommandStep.Builder {
    let _command: string | string[] = [];
    let _env: Record<string, string | number> = {};
    let _plugins: Array<PluginObject | PluginBuilder> = [];
    let _concurrency: number | undefined;
    let _concurrencyGroup: string | undefined;
    let _artifactPaths: string[] = [];
    let _timeoutInMinutes: number | undefined;
    let _softFail: boolean | undefined;
    let _parallelism: number | undefined;
    const keyMixin = KeyMixin.builder();
    const labelMixin = LabelMixin.builder();
    const conditionalMixin = ConditionalMixin.builder();
    const branchesMixin = BranchesMixin.builder();
    const dependenciesMixin = DependenciesMixin.builder();
    return {
      ...(keyMixin as any),
      ...labelMixin,
      ...conditionalMixin,
      ...branchesMixin,
      ...dependenciesMixin,

      command(command) {
        _command = command;
        return this;
      },

      env(env) {
        _env = env;
        return this;
      },

      addEnv(name, value) {
        _env[name] = value;
        return this;
      },

      plugins(plugins) {
        _plugins = plugins;
        return this;
      },

      addPlugin(plugin) {
        _plugins.push(plugin);
        return this;
      },

      concurrency(jobs, group) {
        _concurrency = jobs;
        _concurrencyGroup = group;
        return this;
      },

      artifactPaths(paths) {
        _artifactPaths = paths;
        return this;
      },

      addArtifactPath(path) {
        _artifactPaths.push(path);
        return this;
      },

      timeout(minutes: number) {
        _timeoutInMinutes = minutes;
        return this;
      },

      softFail(fail) {
        _softFail = fail;
        return this;
      },

      parallelism(count) {
        _parallelism = count;
        return this;
      },

      build() {
        const step: CommandStepObject = {
          ...keyMixin.build(),
          ...labelMixin.build(),
          ...conditionalMixin.build(),
          ...branchesMixin.build(),
          ...dependenciesMixin.build(),
        };

        if (Array.isArray(_command) && _command.length > 1) {
          step.commands = _command;
        } else if (Array.isArray(_command) && _command.length === 1) {
          step.command = _command[0];
        } else if (typeof _command === "string") {
          step.command = _command;
        }

        if (Object.keys(_env).length) {
          step.env = _env;
        }

        if (_plugins.length) {
          step.plugins = _plugins.map((plugin) =>
            isBuilder(plugin) ? plugin.build() : plugin
          );
        }

        if (_artifactPaths.length) {
          step.artifact_paths = _artifactPaths;
        }

        if (_timeoutInMinutes) {
          step.timeout_in_minutes = _timeoutInMinutes;
        }

        if (_softFail) {
          step.soft_fail = _softFail;
        }

        if (_concurrency) {
          step.concurrency = _concurrency;
        }
        if (_concurrencyGroup) {
          step.concurrency_group = _concurrencyGroup;
        }

        if (_parallelism) {
          step.parallelism = _parallelism;
        }

        return step;
      },
    };
  }
}
