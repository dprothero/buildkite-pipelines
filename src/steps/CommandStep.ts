import { CommandStepObject } from "./CommandStepObject";
import { PluginBuilder } from "./PluginBuilder";
import { buildKey, configureKey, KeyMixinOptions, mixinKey } from "./mixinKey";
import { StepBuilder } from "./StepBuilder";
import { buildLabel, configureLabel, LabelMixinOptions, mixinLabel } from "./mixinLabel";
import { buildDependency, configureDependency, DependenciesMixinOptions, mixinDependencies } from "./mixinDependencies";
import { BranchesMixinOptions, buildBranches, configureBranches, mixinBranches } from "./mixinBranches";
import { buildConditional, ConditionalMixinOptions, configureConditional, mixinConditional } from "./mixinConditional";

export interface CommandStepOptions extends KeyMixinOptions, LabelMixinOptions, DependenciesMixinOptions, ConditionalMixinOptions, BranchesMixinOptions {
  command?: string | string[]
  if?: string;
  branches?: string[];
  plugins?: PluginBuilder[]

}

export const xCommandStep = class CommandStep implements StepBuilder {
  #dependsOn: (StepBuilder &  {key?: string})[] = [];
  #branches: string[] = [];
  #commands: string[]
  #plugins: PluginBuilder[] = []
  #artifactPaths: string[] = []

  public constructor(command: string | string[] | CommandStepOptions) {
    if (Array.isArray(command)) {
      this.#commands = command 
    } else if (typeof command === 'string') {
      this.#commands = [command]
    } else {
      if (Array.isArray(command.command)) {
        this.#commands = command.command
      } else if (typeof command.command === 'string') {
        this.#commands = [command.command]
      } else {
        this.#commands = []
      }
      this.label = command.label
      this.#plugins = command.plugins ?? []
    }
  }

  public get dependsOn(): StepBuilder[] {
    return this.#dependsOn;
  }

  public dependOn(step: StepBuilder & {key?: string}): this {
    this.#dependsOn?.push(step);
    return this;
  }

  public get branches(): string[] {
    return this.#branches ?? [];
  }

  public addBranch(branch: string): this {
    this.#branches.push(branch);
    return this;
  }

  public build(): CommandStepObject {
    return {
      key: this.#key,
      depends_on: this.#dependsOn.length 
        ? this.#dependsOn.map(step => step.key).filter((key): key is string => !!key)
        : undefined, 
      label: this.#label,
      command: this.#commands[0] ? this.#commands[0] : undefined,
      commands: this.#commands.length > 1 ? this.#commands : undefined,
      plugins: this.#plugins.length ? this.#plugins.map((builder) => builder.build()) : undefined
    }
  }
}

export class CommandStep extends mixinKey(mixinLabel(mixinDependencies(mixinConditional(mixinBranches(class {}))))) implements StepBuilder{
  constructor(options: CommandStepOptions = {}) {
    super()
    configureKey(this, options)
    configureLabel(this, options)
    configureDependency(this, options)
    configureConditional(this, options)
    configureBranches(this, options)
  }

  build(): CommandStepObject {
    return {
      ...buildKey(this),
      ...buildLabel(this),
      ...buildDependency(this),
      ...buildConditional(this),
      ...buildBranches(this),
    }
  }
}

const command = new CommandStep({})
