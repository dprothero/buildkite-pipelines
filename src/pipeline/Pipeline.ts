import { PipelineObject } from "./PipelineObject";
import { StepBuilder } from "./steps/StepBuilder";
import { StepObject } from "./steps/StepObject";

function isBuilder(step: StepObject|StepBuilder): step is StepBuilder {
  return typeof (step as any).build == 'function'
}

export interface PipelineOptions {
  steps?: Array<StepObject | StepBuilder>
}

export class Pipeline {
  #steps: Array<StepObject | StepBuilder>;

  constructor(options: PipelineOptions) {
    this.#steps = options.steps ?? []
  }

  build(): PipelineObject {
    return {
      steps: this.#steps.map(step => {
        if (isBuilder(step)) {
          return step.build()
        } else {
          return step;
        }
      })
    }
  }
}