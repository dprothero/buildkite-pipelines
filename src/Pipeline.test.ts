import {Pipeline} from './Pipeline'
import { CommandStep } from './CommandStep'
import { DockerPlugin } from '.'

describe('Pipeline', () => {
  test('build', () => {

    const pipeline = new Pipeline({
      steps: [
        new CommandStep('yarn run lint'),

        new CommandStep([
          'cd dir',
          'yarn run build'
        ]),

        new CommandStep({
          command: ['']
        })
      ]
    })
    
    expect(pipeline.build()).toMatchSnapshot()
  })


  test('checks', () => {

    const typecheck = new CommandStep({
      label: ':typescript: check:workspaces (back-end)',
      branches: [
        'master',
        'preview',
        'production',
      ],
      command: [
        "set -e",
        "cd back-end",
        "yarn install --frozen-lockfile --development",
        "yarn run setup",
        "yarn run check:workspaces"
      ],
      plugins: [
        new DockerPlugin()
      ]
    })

    const pipeline = new Pipeline({
      steps: [
        typecheck,

        new CommandStep([
          'cd dir',
          'yarn run build'
        ]),

        new CommandStep({
          command: ['']
        })
      ]
    })
    
    expect(pipeline.build()).toMatchSnapshot() 
  })
})