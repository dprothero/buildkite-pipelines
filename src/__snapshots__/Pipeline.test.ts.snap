// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline build 1`] = `
Object {
  "steps": Array [
    Object {
      "commands": Array [
        "yarn run lint",
      ],
      "plugins": Array [],
    },
    Object {
      "commands": Array [
        "cd dir",
        "yarn run build",
      ],
      "plugins": Array [],
    },
    Object {
      "commands": Array [
        "",
      ],
      "plugins": Array [],
    },
  ],
}
`;

exports[`Pipeline checks 1`] = `
Object {
  "steps": Array [
    Object {
      "commands": Array [
        "set -e",
        "cd back-end",
        "yarn install --frozen-lockfile --development",
        "yarn run setup",
        "yarn run check:workspaces",
      ],
      "plugins": Array [],
    },
    Object {
      "commands": Array [
        "cd dir",
        "yarn run build",
      ],
      "plugins": Array [],
    },
    Object {
      "commands": Array [
        "",
      ],
      "plugins": Array [],
    },
  ],
}
`;
